<a href="https://colab.research.google.com/github/Viveksapkale22/saiket_intenship_task-s/blob/main/Saiket_system_intenship_Task_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data science Inenship Work .

**Project Name**: Customer Churn Analysis and Prediction

The project aims to analyze customer churn in a
telecommunications company and develop
predictive models to identify at-risk customers. The
ultimate goal is to provide actionable insights and
recommendations to reduce churn and improve
customer retention.


### Taks 1: Data Preparation

1) Data loading  : use the pandas function

In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler


df = pd.read_csv("/content/drive/MyDrive/Sem-8 work/Intenship_Work/saiket_intenship_selected_1monthvirtualinten/Telco_Customer_Churn_Dataset  (3).csv")
df.head()


customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

2) data exploration : use panda's various function for it.


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [ ]:
df.describe()

SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000

3) Handling missing values: use pandas and numpy to find missing values

- I got to know issue when i see the TotalCharges      7043 non-null   object
- Check the Tenure so it is zero
-  Handling missing by removing it.


In [ ]:
print("Tenure zero:", (df["tenure"] == 0).sum())
print("TotalCharges empty:", (df["TotalCharges"] == ' ').sum())


Tenure zero: 11
TotalCharges empty: 11


In [ ]:
df["TotalCharges"] = df["TotalCharges"].replace(' ', np.nan)
df["tenure"] = pd.to_numeric(df["tenure"], errors="coerce")
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors="coerce")
df.loc[df["tenure"] == 0, "tenure"] = np.nan

print("TotalCharges nan:", df["TotalCharges"].isna().sum())
print("tenure nan:", df["tenure"].isna().sum())

TotalCharges nan: 11
tenure nan: 11


In [ ]:

df.dropna(subset=["tenure", "TotalCharges"], inplace=True)
print("TotalCharges nan:", df["TotalCharges"].isna().sum())
print("tenure nan:", df["tenure"].isna().sum())

TotalCharges nan: 0
tenure nan: 0


In [ ]:
# Handling missing values

ndf = df.isnull().sum()
print(ndf)

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64


4) Data preprocessing : check all unique value and data type
- drop id column because it's no need now.

In [ ]:

col = ['customerID', 'gender',  'SeniorCitizen', 'Partner', 'Dependents',  'tenure', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'DeviceProtection',  'TechSupport', 'StreamingTV', 'StreamingMovies',  'Contract', 'PaperlessBilling',  'PaymentMethod', 'MonthlyCharges' , 'TotalCharges', 'Churn']
#or you can do this
#col = df[1:1:1]     #[start : stop : step]
#col = df.iloc[1:1,0:]    #df.iloc[row_start:row_end, column_start:column_end]
#col = df.columns
for c in col:
 print(c," = ", df[c].unique())


customerID  =  ['7590-VHVEG' '5575-GNVDE' '3668-QPYBK' ... '4801-JZAZL' '8361-LTMKD'
 '3186-AJIEK']
gender  =  ['Female' 'Male']
SeniorCitizen  =  [0 1]
Partner  =  ['Yes' 'No']
Dependents  =  ['No' 'Yes']
tenure  =  [ 1. 34.  2. 45.  8. 22. 10. 28. 62. 13. 16. 58. 49. 25. 69. 52. 71. 21.
 12. 30. 47. 72. 17. 27.  5. 46. 11. 70. 63. 43. 15. 60. 18. 66.  9.  3.
 31. 50. 64. 56.  7. 42. 35. 48. 29. 65. 38. 68. 32. 55. 37. 36. 41.  6.
  4. 33. 67. 23. 57. 61. 14. 20. 53. 40. 59. 24. 44. 19. 54. 51. 26. 39.]
PhoneService  =  ['No' 'Yes']
MultipleLines  =  ['No phone service' 'No' 'Yes']
InternetService  =  ['DSL' 'Fiber optic' 'No']
OnlineSecurity  =  ['No' 'Yes' 'No internet service']
DeviceProtection  =  ['No' 'Yes' 'No internet service']
TechSupport  =  ['No' 'Yes' 'No internet service']
StreamingTV  =  ['No' 'Yes' 'No internet service']
StreamingMovies  =  ['No' 'Yes' 'No internet service']
Contract  =  ['Month-to-month' 'One year' 'Two year']
PaperlessBilling  =  ['Yes' 'No']
PaymentM

In [ ]:
df = df.drop('customerID', axis=1)

5) Categorical variable encoding
- binary encoding
- onehot encoding

In [ ]:

binary_cols = [
    'gender',
    'Partner',
    'Dependents',
    'PhoneService',
    'PaperlessBilling',
    'Churn'
]
binary_map = {
    'Yes': 1,
    'No': 0,
    'Male': 1,
    'Female': 0
}

for col in binary_cols:
    df[col] = df[col].map(binary_map)

df.head()

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0       0              0        1           0     1.0             0   
1       1              0        0           0    34.0             1   
2       1              0        0           0     2.0             1   
3       1              0        0           0    45.0             0   
4       0              0        0           0     2.0             1   

      MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  No phone service             DSL             No          Yes   
1                No             DSL            Yes           No   
2                No             DSL            Yes          Yes   
3  No phone service             DSL            Yes           No   
4                No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

   PaperlessBilling              PaymentMethod  MonthlyCharges  TotalCharges  \
0                 1           Electronic check           29.85         29.85   
1                 0               Mailed check           56.95       1889.50   
2                 1               Mailed check           53.85        108.15   
3                 0  Bank transfer (automatic)           42.30       1840.75   
4                 1           Electronic check           70.70        151.65   

   Churn  
0      0  
1      0  
2      1  
3      0  
4      1

In [ ]:

onehot_cols = [
    'OnlineBackup',
    'MultipleLines',
    'InternetService',
    'OnlineSecurity',
    'DeviceProtection',
    'TechSupport',
    'StreamingTV',
    'StreamingMovies',
    'Contract',
    'PaymentMethod'
]

# Filter onehot_cols to only include columns that are currently of object or category dtype
# This helps prevent errors if some columns were already converted to numeric by other cells
columns_to_onehot = [col for col in onehot_cols if col in df.columns and df[col].dtype in ['object', 'category']]

df = pd.get_dummies(df, columns=columns_to_onehot, drop_first=True)

df.head()

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0       0              0        1           0     1.0             0   
1       1              0        0           0    34.0             1   
2       1              0        0           0     2.0             1   
3       1              0        0           0    45.0             0   
4       0              0        0           0     2.0             1   

   PaperlessBilling  MonthlyCharges  TotalCharges  Churn  ...  \
0                 1           29.85         29.85      0  ...   
1                 0           56.95       1889.50      0  ...   
2                 1           53.85        108.15      1  ...   
3                 0           42.30       1840.75      0  ...   
4                 1           70.70        151.65      1  ...   

   TechSupport_Yes  StreamingTV_No internet service  StreamingTV_Yes  \
0            False                            False            False   
1            False                            False            False   
2            False                            False            False   
3             True                            False            False   
4            False                            False            False   

   StreamingMovies_No internet service  StreamingMovies_Yes  \
0                                False                False   
1                                False                False   
2                                False                False   
3                                False                False   
4                                False                False   

   Contract_One year  Contract_Two year  \
0              False              False   
1               True              False   
2              False              False   
3               True              False   
4              False              False   

   PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                  False                            True   
1                                  False                           False   
2                                  False                           False   
3                                  False                           False   
4                                  False                            True   

   PaymentMethod_Mailed check  
0                       False  
1                        True  
2                        True  
3                       False  
4                       False  

[5 rows x 31 columns]

In [ ]:

#scaler = StandardScaler()
#cols_to_scale = ['tenure', 'MonthlyCharges', 'TotalCharges']

# Fit and transform the selected columns
#df[cols_to_scale] = scaler.fit_transform(df[cols_to_scale])

In [ ]:
df.head()

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0       0              0        1           0     1.0             0   
1       1              0        0           0    34.0             1   
2       1              0        0           0     2.0             1   
3       1              0        0           0    45.0             0   
4       0              0        0           0     2.0             1   

   PaperlessBilling  MonthlyCharges  TotalCharges  Churn  ...  \
0                 1           29.85         29.85      0  ...   
1                 0           56.95       1889.50      0  ...   
2                 1           53.85        108.15      1  ...   
3                 0           42.30       1840.75      0  ...   
4                 1           70.70        151.65      1  ...   

   TechSupport_Yes  StreamingTV_No internet service  StreamingTV_Yes  \
0            False                            False            False   
1            False                            False            False   
2            False                            False            False   
3             True                            False            False   
4            False                            False            False   

   StreamingMovies_No internet service  StreamingMovies_Yes  \
0                                False                False   
1                                False                False   
2                                False                False   
3                                False                False   
4                                False                False   

   Contract_One year  Contract_Two year  \
0              False              False   
1               True              False   
2              False              False   
3               True              False   
4              False              False   

   PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                  False                            True   
1                                  False                           False   
2                                  False                           False   
3                                  False                           False   
4                                  False                            True   

   PaymentMethod_Mailed check  
0                       False  
1                        True  
2                        True  
3                       False  
4                       False  

[5 rows x 31 columns]

In [ ]:
df.info()


<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7042
Data columns (total 31 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   gender                                 7032 non-null   int64  
 1   SeniorCitizen                          7032 non-null   int64  
 2   Partner                                7032 non-null   int64  
 3   Dependents                             7032 non-null   int64  
 4   tenure                                 7032 non-null   float64
 5   PhoneService                           7032 non-null   int64  
 6   PaperlessBilling                       7032 non-null   int64  
 7   MonthlyCharges                         7032 non-null   float64
 8   TotalCharges                           7032 non-null   float64
 9   Churn                                  7032 non-null   int64  
 10  OnlineBackup_No internet service       7032 non-null   bool   
 11  OnlineBac

5) Dataset splitting : do the data spliting in X y for the future use.

In [ ]:
# Dataset splitting.

X = df.drop('Churn', axis=1)
y = df['Churn']


In [ ]:
import os

path = '/content/drive/MyDrive/Sem-8 work/Intenship_Work/saiket_intenship_selected_1monthvirtualinten'
os.makedirs(path, exist_ok=True)

df.to_csv(f"{path}/cleaned_churn_data.csv", index=False)


In [ ]:
!ls "/content/drive/MyDrive/Sem-8 work/Intenship_Work/saiket_intenship_selected_1monthvirtualinten"


 cleaned_churn_data.csv		'SKS Data Science.pdf'
'Saiket system_intenship_work'	'Telco_Customer_Churn_Dataset  (3).csv'
